glean/src/core/metrics/types/string_list.ts

```diff
@@ -0,0 +1,164 @@
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  payload(): string[] {
+    return this._inner;
+  }
+}
+
+/**
+ * A string list metric.
+ *
+ * This allows appending a string value with arbitrary content to a list.
+ * The list is length-limited to `MAX_LIST_LENGTH`.
+ * Strings are length-limited to `MAX_STRING_LENGTH` bytes.
+ */
+class StringListMetricType extends MetricType {
+  constructor(meta: CommonMetricData) {
+    super("string_list", meta);
+  }
+
+  /**
+   * Sets to the specified string list value.
+   *
+   * # Note
+   *
+   * Truncates the list if it is longer than `MAX_LIST_LENGTH` and logs an error.
+   *
+   * Truncates the value if it is longer than `MAX_STRING_LENGTH` bytes
+   * and logs an error.
+   *
+   * @param value the list of string to set the metric to.
+   */
+  set(value: string[]): void {
+    Context.dispatcher.launch(async () => {
+      if (!this.shouldRecord(Context.uploadEnabled)) {
+        return;
+      }
+      const truncatedList: string[] = [];
+      if (value.length > MAX_LIST_LENGTH) {
+        await Context.errorManager.record(
+          this,
+          ErrorType.InvalidValue,
+          `String list length of ${value.length} exceeds maximum of ${MAX_LIST_LENGTH}.`
+        );
+      }
+
+      for (let i = 0; i < Math.min(value.length, MAX_LIST_LENGTH); ++i) {
+        const truncatedString = await truncateStringAtBoundaryWithError(this, value[i], MAX_STRING_LENGTH);
+        truncatedList.push(truncatedString);
+      }
+      const metric = new StringListMetric(truncatedList);
+      await Context.metricsDatabase.record(this, metric);
+    });
+  }
+
+  /**
+   * Adds a new string `value` to the list.
+   *
+   * # Note
+   *
+   * - If the list is already of length `MAX_LIST_LENGTH`, log an error.
+   * - Truncates the value if it is longer than `MAX_STRING_LENGTH` bytes
+   * and logs an error.
+   *
+   * @param value The string to add.
+   */
+  add(value: string): void {
+    Context.dispatcher.launch(async () => {
+      if (!this.shouldRecord(Context.uploadEnabled)) {
+        return;
+      }
+
+      const truncatedValue = await truncateStringAtBoundaryWithError(this, value, MAX_STRING_LENGTH);
+      let currentLen = -1;
+
+      const transformFn = ((value) => {
+        return (v?: JSONValue): StringListMetric => {
+          let metric: StringListMetric;
+          let result: string[];
+          try {
+            metric = new StringListMetric(v);
+            result = metric.get();
+            currentLen = result.length;
+            if (result.length < MAX_LIST_LENGTH) {
+              result.push(value);
+            }
+          } catch {
+            metric = new StringListMetric([value]);
+            result = [value];
+          }
+          metric.set(result);
+          return metric;
+        };
+      })(truncatedValue);
+
+      await Context.metricsDatabase.transform(this, transformFn);
+
+      if (currentLen == MAX_LIST_LENGTH) {
```

Comment:
```suggestion
      if (currentLen >= MAX_LIST_LENGTH) {
```

Just to be extra sure, hehe.